File: backend/services/analysis.service.js
/**
 * Analyse a completed interview session and return a rich, specific report.
 * Called by session.controller.js after the final answer is submitted.
 *
 * Strategy:
 * 1. Try Ollama batch evaluation for a holistic AI-generated analysis (best quality)
 * 2. If Ollama fails (timeout / error), fall back to a smart aggregation
 *    of local clarity scores and heuristic data — producing real, specific feedback
 *    instead of generic placeholders.
 */

import { batchEvaluateInterview, computeLocalClarityScore } from "./ai.service.js";

// ── Smart fallback builder ──────────────────────────────────────────────────
/**
 * Build a real analysis report purely from local heuristic data.
 * Used when Ollama is unavailable (timeout, error, etc.)
 */
function buildFallbackReport({ answered, domain, role }) {
  // Compute local clarity for each answer
  const clarityData = answered.map((q) => computeLocalClarityScore(q.answerText));
  const avgClarity = clarityData.reduce((s, c) => s + c.clarityScore, 0) / (clarityData.length || 1);

  // Estimate content score from word count and tech terms (rough heuristic)
  const contentEstimates = answered.map((q) => {
    const wc = q.answerText?.trim().split(/\s+/).length ?? 0;
    const techTerms = /\b(function|class|api|database|algorithm|complexity|cache|async|thread|memory|server|client|request|response|component|module|interface|architecture)\b/gi;
    const techCount = (q.answerText?.match(techTerms) || []).length;
    if (wc < 15) return 2;
    if (wc < 30) return 3 + Math.min(2, techCount);
    if (techCount >= 3) return 6 + Math.min(2, Math.floor(techCount / 3));
    return 4;
  });
  const avgContent = contentEstimates.reduce((s, v) => s + v, 0) / (contentEstimates.length || 1);

  // Voice confidence from signals
  const confScores = answered.map((q) => {
    const v = q.confidenceSignals?.voice;
    return v != null ? Math.min(10, Math.max(1, v)) : 5;
  });
  const avgConfidence = confScores.reduce((s, v) => s + v, 0) / (confScores.length || 1);

  const avgOverall = avgContent * 0.5 + avgClarity * 0.3 + avgConfidence * 0.2;

  // Build per-answer evaluations for the transcript
  const perAnswerEvaluations = answered.map((q, i) => {
    const wc = q.answerText?.trim().split(/\s+/).length ?? 0;
    const isShort = wc < 15;
    const content = isShort ? Math.min(contentEstimates[i], 3) : contentEstimates[i];
    const clarity = isShort ? Math.min(clarityData[i].clarityScore, 3) : clarityData[i].clarityScore;
    const conf = isShort ? Math.min(confScores[i], 4) : confScores[i];
    const overall = Math.round((content * 0.5 + clarity * 0.3 + conf * 0.2) * 10) / 10;

    return {
      contentScore: content,
      clarityScore: clarity,
      confidenceScore: conf,
      overallScore: overall,
      strengths: wc >= 50 ? ["Provided a detailed response"] : ["Attempted to answer"],
      improvements: isShort
        ? ["Answer too brief — expand with examples and reasoning"]
        : ["Add more specific technical details"],
    };
  });

  // Build summary
  const perfLabel = avgOverall >= 8 ? "strong" : avgOverall >= 6 ? "solid" : avgOverall >= 4 ? "moderate" : "weak";
  const contentNote = avgContent < 5
    ? "Technical accuracy and depth need significant improvement."
    : avgContent < 7
    ? "Technical content was partially correct but lacked depth in several areas."
    : "Technical content was generally accurate and well-explained.";

  const clarityNote = avgClarity < 5
    ? "Answers were often unclear or poorly structured."
    : avgClarity < 7
    ? "Clarity was adequate but could benefit from more structured responses."
    : "Communication was clear and well-organized.";

  const summary = `Completed ${answered.length} questions in ${role || domain} with an overall score of ${avgOverall.toFixed(1)}/10 — a ${perfLabel} performance. ${contentNote} ${clarityNote}`;

  // Strengths
  const topStrengths = [];
  if (avgClarity >= 6) topStrengths.push("Clear and structured communication throughout the interview");
  if (avgContent >= 6) topStrengths.push("Demonstrated solid technical knowledge in core areas");
  if (avgConfidence >= 6) topStrengths.push("Confident delivery and consistent speaking pace");
  if (topStrengths.length === 0) topStrengths.push("Made an effort to answer all questions presented");

  // Improvements
  const topImprovements = [];
  if (avgContent < 6) topImprovements.push("Deepen technical knowledge — answers lacked precision and factual accuracy");
  if (avgClarity < 6) topImprovements.push("Structure answers more clearly — use concept → example → trade-offs flow");
  if (avgContent < 7) topImprovements.push("Provide concrete, specific examples from real projects");
  const shortAnswers = answered.filter((q) => (q.answerText?.trim().split(/\s+/).length ?? 0) < 30);
  if (shortAnswers.length > 0) {
    topImprovements.push(`Expand answers — ${shortAnswers.length} answer(s) were too brief`);
  }
  if (topImprovements.length === 0) topImprovements.push("Continue practicing to maintain performance");

  // Recommendations
  const recommendations = [];
  if (avgContent < 6) recommendations.push(`Study core ${domain} fundamentals using official documentation and courses`);
  if (avgClarity < 6) recommendations.push("Practice the STAR method and record yourself answering questions");
  if (avgOverall < 7) recommendations.push("Do 2-3 mock interviews per week on platforms like Pramp or Interviewing.io");
  recommendations.push("Review the specific questions where you scored lowest and write out comprehensive answers");

  const level = avgConfidence > 7 ? "High" : avgConfidence >= 5 ? "Medium" : "Low";

  const salaryReadiness = avgOverall >= 7
    ? `With an overall score of ${avgOverall.toFixed(1)}/10, performance is on track for the target salary range.`
    : `An overall score of ${avgOverall.toFixed(1)}/10 suggests further preparation is needed before targeting the expected salary range.`;

  const keyInsight = topImprovements[0]
    ? `Priority focus: ${topImprovements[0]}`
    : "Focus on providing concrete examples and specific technical details in every answer.";

  return {
    perAnswerEvaluations,
    report: {
      summary,
      topStrengths: topStrengths.slice(0, 3),
      topImprovements: topImprovements.slice(0, 3),
      recommendations: recommendations.slice(0, 3),
      confidenceLevel: level,
      salaryReadiness,
      keyInsight,
      scores: {
        content: parseFloat(avgContent.toFixed(2)),
        clarity: parseFloat(avgClarity.toFixed(2)),
        confidence: parseFloat(avgConfidence.toFixed(2)),
        overall: parseFloat(avgOverall.toFixed(2)),
      },
    },
  };
}


// ── Main export ─────────────────────────────────────────────────────────────

export const analyzeInterview = async ({ questions, domain, role, experienceLevel, salaryRange }) => {
  const answered = questions.filter((q) => q.answerText);

  if (answered.length === 0) {
    return buildFallbackReport({ answered: [], domain, role }).report;
  }

  try {
    console.log(`🧠 Starting batch evaluation for ${answered.length} answers...`);

    const result = await batchEvaluateInterview({
      questions: answered,
      domain,
      role,
      experienceLevel,
      salaryRange,
    });

    // Attach per-answer evaluations back onto the question objects
    // so they appear in the transcript
    answered.forEach((q, i) => {
      if (result.perAnswerEvaluations[i]) {
        q.evaluation = result.perAnswerEvaluations[i];
      }
    });

    // Filter out generic AI fallback strings
    const isGeneric = (s) =>
      !s ||
      s.toLowerCase().includes("answer was submitted") ||
      s.toLowerCase().includes("ai evaluation temporarily") ||
      s.toLowerCase().includes("please retry");

    const cleanList = (arr, fallbackArr) =>
      Array.isArray(arr) && arr.filter((s) => !isGeneric(s)).length >= 2
        ? arr.filter((s) => !isGeneric(s))
        : fallbackArr;

    const fallback = buildFallbackReport({ answered, domain, role });

    return {
      ...result.report,
      topStrengths: cleanList(result.report.topStrengths, fallback.report.topStrengths),
      topImprovements: cleanList(result.report.topImprovements, fallback.report.topImprovements),
      recommendations: Array.isArray(result.report.recommendations) && result.report.recommendations.length > 0
        ? result.report.recommendations
        : fallback.report.recommendations,
    };
  } catch (err) {
    console.error("analyzeInterview Ollama batch eval failed — using smart fallback:", err.message);

    // Smart fallback: real feedback from local heuristic data
    const fallback = buildFallbackReport({ answered, domain, role });

    // Attach fallback per-answer evaluations
    answered.forEach((q, i) => {
      if (fallback.perAnswerEvaluations[i]) {
        q.evaluation = fallback.perAnswerEvaluations[i];
      }
    });

    return fallback.report;
  }
};

File: backend_django/services/analysis_service.py
"""
analysis_service.py
Python port of analysis.service.js.
Orchestrates interview analysis using Ollama batch eval (primary) + heuristic fallback.
"""
import logging
from .ai_service import batch_evaluate_interview, compute_local_clarity_score

logger = logging.getLogger(__name__)


def _build_fallback_report(answered: list, domain: str, role=None) -> dict:
    """
    Build a real analysis report purely from local heuristic data.
    Used when Ollama is unavailable.
    Equivalent to buildFallbackReport() in analysis.service.js.
    """
    clarity_data = [compute_local_clarity_score(q["answerText"]) for q in answered]
    avg_clarity = sum(c["clarityScore"] for c in clarity_data) / (len(clarity_data) or 1)

    tech_pattern = (
        r"\b(function|class|api|database|algorithm|complexity|cache|async|thread|memory|"
        r"server|client|request|response|component|module|interface|architecture)\b"
    )

    import re
    def _content_estimate(q: dict) -> float:
        wc = len(q["answerText"].split()) if q.get("answerText") else 0
        tech_count = len(re.findall(tech_pattern, q.get("answerText", ""), re.IGNORECASE))
        if wc < 15:  return 2
        if wc < 30:  return 3 + min(2, tech_count)
        if tech_count >= 3: return 6 + min(2, tech_count // 3)
        return 4

    content_estimates = [_content_estimate(q) for q in answered]
    avg_content = sum(content_estimates) / (len(content_estimates) or 1)

    conf_scores = [min(10, max(1, q.get("confidenceSignals", {}).get("voice", 5) or 5)) for q in answered]
    avg_confidence = sum(conf_scores) / (len(conf_scores) or 1)
    avg_overall = avg_content * 0.5 + avg_clarity * 0.3 + avg_confidence * 0.2

    per_answer_evaluations = []
    for i, q in enumerate(answered):
        wc = len(q["answerText"].split()) if q.get("answerText") else 0
        is_short = wc < 15
        content = min(content_estimates[i], 3) if is_short else content_estimates[i]
        clarity  = min(clarity_data[i]["clarityScore"], 3) if is_short else clarity_data[i]["clarityScore"]
        conf     = min(conf_scores[i], 4) if is_short else conf_scores[i]
        overall  = round((content * 0.5 + clarity * 0.3 + conf * 0.2) * 10) / 10
        per_answer_evaluations.append({
            "contentScore":    content,
            "clarityScore":    clarity,
            "confidenceScore": conf,
            "overallScore":    overall,
            "strengths":    ["Provided a detailed response"] if wc >= 50 else ["Attempted to answer"],
            "improvements": ["Answer too brief — expand with examples and reasoning"] if is_short else ["Add more specific technical details"],
        })

    perf_label = "strong" if avg_overall >= 8 else ("solid" if avg_overall >= 6 else ("moderate" if avg_overall >= 4 else "weak"))
    content_note = (
        "Technical accuracy and depth need significant improvement." if avg_content < 5
        else "Technical content was partially correct but lacked depth in several areas." if avg_content < 7
        else "Technical content was generally accurate and well-explained."
    )
    clarity_note = (
        "Answers were often unclear or poorly structured." if avg_clarity < 5
        else "Clarity was adequate but could benefit from more structured responses." if avg_clarity < 7
        else "Communication was clear and well-organized."
    )

    summary = (
        f"Completed {len(answered)} questions in {role or domain} with an overall score of {avg_overall:.1f}/10 "
        f"— a {perf_label} performance. {content_note} {clarity_note}"
    )

    top_strengths = []
    if avg_clarity    >= 6: top_strengths.append("Clear and structured communication throughout the interview")
    if avg_content    >= 6: top_strengths.append("Demonstrated solid technical knowledge in core areas")
    if avg_confidence >= 6: top_strengths.append("Confident delivery and consistent speaking pace")
    if not top_strengths:   top_strengths.append("Made an effort to answer all questions presented")

    top_improvements = []
    if avg_content < 6: top_improvements.append("Deepen technical knowledge — answers lacked precision and factual accuracy")
    if avg_clarity < 6: top_improvements.append("Structure answers more clearly — use concept → example → trade-offs flow")
    if avg_content < 7: top_improvements.append("Provide concrete, specific examples from real projects")
    short_answers = [q for q in answered if len((q.get("answerText") or "").split()) < 30]
    if short_answers:
        top_improvements.append(f"Expand answers — {len(short_answers)} answer(s) were too brief")
    if not top_improvements:
        top_improvements.append("Continue practicing to maintain performance")

    recommendations = []
    if avg_content < 6: recommendations.append(f"Study core {domain} fundamentals using official documentation and courses")
    if avg_clarity < 6: recommendations.append("Practice the STAR method and record yourself answering questions")
    if avg_overall < 7: recommendations.append("Do 2-3 mock interviews per week on platforms like Pramp or Interviewing.io")
    recommendations.append("Review the specific questions where you scored lowest and write out comprehensive answers")

    level = "High" if avg_confidence > 7 else ("Medium" if avg_confidence >= 5 else "Low")
    salary_readiness = (
        f"With an overall score of {avg_overall:.1f}/10, performance is on track for the target salary range."
        if avg_overall >= 7
        else f"An overall score of {avg_overall:.1f}/10 suggests further preparation is needed before targeting the expected salary range."
    )
    key_insight = f"Priority focus: {top_improvements[0]}" if top_improvements else "Focus on providing concrete examples and specific technical details in every answer."

    return {
        "perAnswerEvaluations": per_answer_evaluations,
        "report": {
            "summary":         summary,
            "topStrengths":    top_strengths[:3],
            "topImprovements": top_improvements[:3],
            "recommendations": recommendations[:3],
            "confidenceLevel": level,
            "salaryReadiness": salary_readiness,
            "keyInsight":      key_insight,
            "scores": {
                "content":    round(avg_content, 2),
                "clarity":    round(avg_clarity, 2),
                "confidence": round(avg_confidence, 2),
                "overall":    round(avg_overall, 2),
            },
        },
    }


def analyze_interview(*, questions: list, domain: str, role=None, experience_level=None, salary_range=None) -> dict:
    """
    Analyse a completed interview session.
    Tries Ollama batch evaluation first; falls back to heuristic analysis.
    Equivalent to analyzeInterview() in analysis.service.js.
    """
    answered = [q for q in questions if q.get("answerText")]

    if not answered:
        return _build_fallback_report([], domain, role)["report"]

    def _is_generic(s: str) -> bool:
        if not s:
            return True
        sl = s.lower()
        return (
            "answer was submitted" in sl
            or "ai evaluation temporarily" in sl
            or "please retry" in sl
        )

    def _clean_list(arr, fallback):
        if isinstance(arr, list) and len([s for s in arr if not _is_generic(s)]) >= 2:
            return [s for s in arr if not _is_generic(s)]
        return fallback

    try:
        logger.info("🧠 Starting batch evaluation for %d answers...", len(answered))
        result = batch_evaluate_interview(
            questions=answered,
            domain=domain,
            role=role,
            experience_level=experience_level,
            salary_range=salary_range,
        )

        # Attach per-answer evaluations back onto question objects
        for i, q in enumerate(answered):
            if i < len(result["perAnswerEvaluations"]):
                q["evaluation"] = result["perAnswerEvaluations"][i]

        fallback = _build_fallback_report(answered, domain, role)
        report = result["report"]

        return {
            **report,
            "topStrengths":    _clean_list(report.get("topStrengths"),    fallback["report"]["topStrengths"]),
            "topImprovements": _clean_list(report.get("topImprovements"), fallback["report"]["topImprovements"]),
            "recommendations": (
                report.get("recommendations")
                if isinstance(report.get("recommendations"), list) and report["recommendations"]
                else fallback["report"]["recommendations"]
            ),
        }

    except Exception as exc:
        logger.error("analyzeInterview Ollama batch eval failed — using smart fallback: %s", exc)
        fallback = _build_fallback_report(answered, domain, role)
        for i, q in enumerate(answered):
            if i < len(fallback["perAnswerEvaluations"]):
                q["evaluation"] = fallback["perAnswerEvaluations"][i]
        return fallback["report"]
